segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/metadata/BaseMetadataStore.java

```diff
@@ -299,110 +487,169 @@ public void abort(MetadataTransaction txn) throws StorageMetadataException {
-        TransactionData dataFromBuffer = null;
         if (null == key) {
-            return null;
+            return CompletableFuture.completedFuture(null);
         }
-        StorageMetadata retValue = null;
 
         Map<String, TransactionData> txnData = txn.getData();
+
+        // Record is found in transaction data itself.
         TransactionData data = txnData.get(key);
+        if (null != data) {
+            return CompletableFuture.completedFuture(data.getValue());
+        }
 
-        // Search in the buffer.
-        if (null == data) {
-            synchronized (lock) {
-                dataFromBuffer = bufferedTxnData.get(key);
-            }
-            // If we did not find in buffer then load it from store
-            if (null == dataFromBuffer) {
-                // NOTE: This call to read MUST be outside the lock, it is most likely cause re-entry.
-                loadFromStore(key);
-                dataFromBuffer = bufferedTxnData.get(key);
-                Preconditions.checkState(null != dataFromBuffer);
-            }
+        // Prevent the key from getting evicted.
+        addToActiveKeySet(key);
+
+        // Try to find it in buffer. Access buffer using reader lock.
+        val readLock = scheduler.getReadLock(new String[]{key});
+        return readLock.lock()
+                .thenApplyAsync(v -> bufferedTxnData.get(key), executor)
+                .thenApplyAsync(dataFromBuffer -> {
+                    if (dataFromBuffer != null) {
+                        // Make sure it is a deep copy.
+                        val retValue = dataFromBuffer.getValue();
+                        if (null != retValue) {
+                            return retValue.deepCopy();
+                        }
+                        return null;
+                    }
+                    return null;
+                }, executor)
+                .whenCompleteAsync((v, ex) -> readLock.unlock(), executor)
+                .thenComposeAsync(retValue -> {
+                    if (retValue != null) {
+                        return CompletableFuture.completedFuture(retValue);
+                    }
+                    // We did not find it in the buffer either.
+                    // Try to find it in store.
+                    return loadFromStore(key, false)
+                            .thenApplyAsync(TransactionData::getValue, executor);
+                }, executor)
+                .whenCompleteAsync((v, ex) -> removeFromActiveKeySet(key), executor);
+    }
 
-            if (null != dataFromBuffer && null != dataFromBuffer.getValue()) {
-                // Make copy.
-                data = dataFromBuffer.toBuilder()
-                        .key(key)
-                        .value(dataFromBuffer.getValue().deepCopy())
-                        .build();
-                txnData.put(key, data);
+    private void removeFromActiveKeySet(String key) {
+        activeKeys.remove(key);
```

Comment:
You are not holding a lock here while accessing `activeKeys` while below you do. Please be consistent (and correct).

Reply:
activeKeys is concurrent multi hash - so addition/deleting is safe in this context.
added comment explaining why not having lock here is safe.